README.md
# SimpleDOI [![Build Status](https://travis-ci.org/UMNLibraries/simpledoi-ruby.svg?branch=master)](https://travis-ci.org/UMNLibraries/simpledoi-ruby)
A library for locating DOIs and retrieving metadata about their targets


## Basic Usage
### Retrieve the target resource URL from a known DOI

```ruby
require 'simple_doi'

# Instantiate a DOI object based on a DOI string
doi = SimpleDOI::DOI.new '10.1000/182'

# Override "curb" or "net/http" as the resolver backend
# Defaults to 'curb' if Curl is already in use (curb already required), otherwise Net/HTTP
doi.backend = 'curb'

# Resolve the target URL
puts doi.target_url
# Prints: "http://www.doi.org/hb.html"
```

### Extract one or more DOIs from a larger string, such as a URL
```ruby
# An HTTP backend must be loaded
require 'curb'
# alternatively
# require 'net/http'

require 'simple_doi'

input = 'http://dx.doi.org/10.1000/182'

# SimpleDOI#extract returns an array of DOI objects
dois = SimpleDOI.extract input
puts dois.first.inspect
# Prints: [#<SimpleDOI::DOI:0x007f6316bf14d8 @doi="10.1000/182", @target_url=nil, @backend=curb>]

# Note the @target_url has not yet been resolved
```

### Retrieve metadata about the target resource
Via content negotiation `Accept:` headers, JSON or XML metadata can be
retrieved.  The `SimpleDOI::MetadataParser` module provides _very basic_ parsing
for `application/citeproc+json` and `application/unixref+xml` content types.
Review the source files in `lib/metadata_parser/*.rb` to see available metadata
methods.

Note that not all targets may be able to supply your requested format. Content
negotiation allows you to request multiple types in order of preference, and the
`DOI#lookup` method will return the first it could retrieve while setting its
`response_content_type` property accordingly.

```ruby
# Retrieve JSON metadata (default format)
doi = SimpleDOI::DOI.new('10.1111/j.1475-3995.1998.tb00130.x')

# Call lookup() with no format specified for JSON
json = doi.lookup
puts doi.response_content_type
# "application/vnd.citationstyles.csl+json"

# Retrieve Unixref XML
doi = SimpleDOI::DOI.new('10.1111/j.1475-3995.1998.tb00130.x')

# Call lookup() and specify the
xml = doi.lookup SimpleDOI::UNIXREF_XML
puts doi.response_content_type
# "application/vnd.crossref.unixref+xml"

# Call lookup() and prefer XML, but fallback to JSON if unavailable
response = doi.lookup [SimpleDOI::UNIXREF_XML, SimpleDOI::CITEPROC_JSON]
# Check the response_content_type to verify you got what you wanted
if doi.response_content_type == SimpleDOI::UNIXREF_XML
  # Handle it as XML...
else
  # Handle it as JSON
end

```

Valid format arguments for `lookup` are `SimpleDOI::UNIXREF_XML,
SimpleDOI::CITEPROC_JSON`

### Parse returned metadata
The metadata format parsers included with this library are not comprehensive.
They present a common, simple interface to retrieve certain metadata components
we have needed in the past, but if you have more complex requirements you can
always pass the response body to a general purpose interface like `JSON` or
`Nokogiri`.

`SimpleDOI::MetadataParser::CiteprocJSONParser` and
`SimpleDOI::MetadataParser::UnixrefXMLParser` accept the `body` attribute set on
a `DOI` object after lookup and provide wrapper methods to retrieve certain
important bibliographic attributes.

```ruby
# Lookup an identifier, requesting Unixref XML
doi = SimpleDOI::DOI.new('10.1111/j.1475-3995.1998.tb00130.x')
doi.lookup SimpleDOI::UNIXREF_XML

# Pass the body to a UnixrefXMLParser initializer
parser = SimpleDOI::MetadataParser::UnixrefXMLParser.new(doi.body)

# Is it a book/mongraph?
parser.book?
# false

# Is it a general journal/serial (not article)?
parser.journal?
# false

# Is it a journal/serial article
parser.journal_article?
# true

# Is it a conference proceeding?
parser.conference_proceeding?
# false

# Get the ISSN
parser.issn
# "0969-6016"

# Get the journal title
parser.journal_title
# "International Transactions in Operational Research"

# Get the article title
parser.article_title
# "Passenger Choice Analysis for Seat Capacity Control...."

# Get some publication information
"Vol: #{parser.volume} Issue: #{parser.issue} Pages: #{parser.pagination} Published: #{parser.publication_date}"
# "Vol: 5 Issue: 6 Pages: 471-486"
# "Published: 1998-11-01"

# Get the authors/editors/contributors
parser.contributors
# "[#<struct SimpleDOI::MetadataParser::Parser::Contributor given_name=\"S-E.\", surname=\"Andersson\", contributor_role=\"author\", sequence=1>]"

# If multiple contributor_role values, shorthand as:
parser.authors
parser.editors

# Bundle up the whole thing as a Hash
parser.to_h
# {:book_title=>nil, :book_series_title=>nil, :isbn=>nil, :eisbn=>nil, :journal_title=>"International Transactions in Operational Research", :journal_isoabbrev_title=>"Int Trans Operational Res", :issn=>"0969-6016", :eissn=>"1475-3995", :article_title=>"Passenger Choice Analysis for Seat Capacity Control: A Pilot Project in Scandinavian Airlines", :conference_title=>nil, :contributors=>[{:given_name=>"S-E.", :surname=>"Andersson", :contributor_role=>"author", :sequence=>1}], :doi=>"10.1111/j.1475-3995.1998.tb00130.x", :url=>"http://doi.wiley.com/10.1111/j.1475-3995.1998.tb00130.x", :publisher=>nil, :volume=>"5", :issue=>"6", :pagination=>"471-486", :publication_date=>#<Date: 1998-11-01 ((2451119j,0s,0n),+0s,2299161j)>}
```

All currently implemented methods are listed in
[`lib/metadata_parser.rb`](lib/metadata_parser.rb)

## Backends
SimpleDOI supports either Net/HTTP or cURL (via Curb) as its resolver backend.
On instantiation, it will attempt to detect if Curb is present and use it,
falling back to Net/HTTP, and raising an error if neither is available.

The purpose of this is to avoid having to `require` curl when Net/HTTP is
already in use. It comes at the cost of not declaring the Curb dependency and is
maybe not helpful. It could be removed in the future, depending on one or the
other.

If neither `curb` nor `net/http` has been required previously,
`SimpleDOI::DOI.new` will raise an error.

## Supported DOIs
The University of Minnesota Libraries has collected many diverse thousands of
DOIs, mostly from URLs passing through proxy servers. This library attempts to
support all of those. They range from very simple to extremely complex.

- `10.1023/B:SOVI.0000043002.02424.ca`
- `10.1002/(SICI)1099-1050(199609)5:5!!447::AID-HEC220!!3.0.CO;2-#`

Certain common patterns are used to extract additional ISSN identifiers
from the DOIs. **Note**: These are strictly pattern matches from the DOI string,
and therefore not guaranteed to actually _be_ valid ISSNs. In our experience,
and for our purposes they are sufficient.

- `10.1111/j.1475-3995.1998.tb00130.x` will yield `1475-3995` as its ISSN

For many examples of supported strings, see
[`tests/fixtures/good-doi.txt`](tests/fixtures/good-doi.txt).

## Building the gem
```shell
# Build a .gem of the current version
$ gem build simple_doi.gemspec

# Perform a local install
$ gem install simple_doi-x.x.x.gem
```

## Testing
```shell
$ bundle exec rake test
```

## Known issues


CHANGELOG.md
# CHANGELOG
1.1.0:
* Support for URL encoded DOI strings with `SimpleDOI#extract`
* Changes dx.doi.org base redirector to use https://

1.0.0:
 * `SimpleDOI::MetadataParser#journal?` behavior has changed and
   `SimpleDOI::MetadataParser#journal_article?` was added.  Previously, calls to
   `journal?` would return true for DOI's pointing to articles or DOI's pointing
   to the journal itself rather than a specific article. As of 1.0.0,
   `.journal?` returns `true` for the journal record only.  Use
   `journal_article?` to determine if the metadata represents a single article
   within a containing journal/publication.


tests/test_simple_doi.rb
require 'minitest/autorun'
require 'webmock'
require 'webmock/minitest'
require_relative '../lib/simple_doi/simple_doi'

module SimpleDOI
  module Test
    class SimpleDOITest < MiniTest::Test
      WebMock.disable_net_connect!
      def setup
      end
      def teardown
      end

      # Test a long list of known good DOIs
      def test_valid_doi
        File.foreach File.dirname(__FILE__) + '/fixtures/good-doi.txt' do |doi|
          assert SimpleDOI.valid?(doi), "#{doi} should be considered a valid DOI"
        end
      end

      # Test a list of known faulty DOIs
      def test_invalid_doi
        File.foreach File.dirname(__FILE__) + '/fixtures/bad-doi.txt' do |doi|
          refute SimpleDOI.valid?(doi), "#{doi} should not be considered a valid DOI"
        end
      end

      def test_initialize
        # Valid DOI, creates object
        doi = '10.1234/abcdefg'
        doi_obj = SimpleDOI::DOI.new doi
        assert_instance_of SimpleDOI::DOI, doi_obj, '#{doi} should result in a valid SimpleDOI::DOI object'

        # Invalid DOI throws exception
        assert_raises ArgumentError do
          SimpleDOI::DOI.new('not a DOI')
        end

        # dx.doi.org info stripped
        doi = 'https://dx.doi.org/10.1234/abcdefg?nosfx=y'
        doi_obj = SimpleDOI::DOI.new doi
        assert doi_obj.to_s == doi, 'dx.doi.org URL info should be stripped'
      end

      def test_backend
        doi = SimpleDOI::DOI.new '10.1234/abcdefg'
        # Should work fine
        assert_equal 'curb', (doi.backend = 'curb')
        assert_equal 'net/http', (doi.backend = 'net/http')

        # Should fail invalid
        assert_raises ArgumentError do
          doi.backend = 'UNSUPPORTED'
        end
      end

      def test_prefix_suffix
        doi = SimpleDOI::DOI.new '10.1234/abcdefg.123456'
        assert_equal '10.1234', doi.prefix

        doi = SimpleDOI::DOI.new '10.1234.56/abcdefg.123456'
        assert_equal '10.1234.56', doi.prefix

        doi = SimpleDOI::DOI.new '10.1234.56/abcdefg.123456'
        assert_equal 'abcdefg.123456', doi.suffix

        # Multiple /, all in the suffix
        doi = SimpleDOI::DOI.new '10.1234/abcdefg/123456/xyz.99'
        assert_equal 'abcdefg/123456/xyz.99', doi.suffix
      end

      def test_extract_doi
        File.foreach File.dirname(__FILE__) + '/fixtures/extract-single.txt' do |line|
          expected, haystack = line.split ' ', 2
          extracted = SimpleDOI.extract haystack
          assert_kind_of SimpleDOI::DOI, extracted.first, 'The extracted object should be a SimpleDOI::DOI object'
          assert_equal expected, extracted.first.doi, "A DOI #{expected} should be found in the string and returned"
        end
      end

      def test_extract_multiple_doi
        # Multi tests repeat the same string twice
        # so each should return 2 DOIs and both should match expected
        File.foreach File.dirname(__FILE__) + '/fixtures/extract-multi.txt' do |line|
          expected, haystack = line.split ' ', 2
          found = SimpleDOI.extract haystack
          assert_equal 2, found.count, "Two DOIs should be found in the haystack string '#{haystack}'"
          assert_equal expected, found.first.doi
          assert_equal expected, found.last.doi
        end
      end

      def test_issn
        good_issn = [
          ['1234-1234', '10.1234/j.1234-1234.abcd'],
          ['1234-1234', '10.1234/ISSN.1234-1234.abcd'],
          ['1234-123X', '10.1234/J.1234-123X.abcd'],
          ['1234-123X', '10.1234/ISSN.1234-123X.abcd'],
          ['1234-123X', '10.1234/ISSN.1234-123X.abcd'],
          ['1234-1234', '10.1234/1234-1234(STUFF)abcd'],
          ['1234-123X', '10.1234/1234-123X(STUFF)abcd'],
          ['1234-123X', '10.1234/1234-123X.abcd'],
          ['1234-123X', '10.1234/S1234-123X(STUFF).abcd'],
          ['1234-123X', '10.1234/(ISSN)1234-123X.STUFF.abcd']
        ]
        bad_issn = [
          '10.1234/j.12345-1234.abc',
          '10.1234/no.issn-here',
          '10.1234/j.1234-123Z.abc',
          '10.1234/issn.1234-123z.abc',
          '10.1234/1234-1234noboundary.abc',
          '10.1234/1234-123tooshort.abc',
          '10.1234/XYZS1234-123X(STUFF).abcd'
        ]
        good_issn.each do |pair|
          doi = SimpleDOI::DOI.new pair[1]
          assert_equal pair[0], doi.issn, "The ISSN #{pair[0]} should be returned"
        end
        bad_issn.each do |doi_str|
          doi = SimpleDOI::DOI.new doi_str
          assert_nil doi.issn, "No ISSN string should be found in #{doi_str}"
        end
      end

      def test_url
        doi = SimpleDOI::DOI.new '10.1234/abcd.321'
        assert_equal 'https://dx.doi.org/10.1234%2Fabcd.321', doi.url, 'A full CGI escaped URL should be returned'
      end

      def test_lookup_json
        # Mock a set of returns first to dx.doi.org with a location redirect,
        # then to a target resource returning JSON
        expected = { 'prop1' => 'property', 'prop2' => 'property' }

        stub_request(:get, /https:\/\/dx.doi.org\/(.+)json/)
          .to_return(body: 'This is redirecting', status: 301, headers: { 'location' => 'http://example.com/doijson' })
        stub_request(:get, 'http://example.com/doijson')
          .to_return(body: expected.to_json, status: 200, headers: { 'content-type' => 'application/citeproc+json' })

        # Lookup with curb
        doi = SimpleDOI::DOI.new '10.1234/abcd.json'
        doi.backend = 'curb'
        hash = doi.lookup_json
        assert_equal expected, hash

        # Lookup with Net::HTTP
        doi = SimpleDOI::DOI.new '10.1234/abcd.json'
        doi.backend = 'net/http'
        hash = doi.lookup_json
        assert_equal expected, hash
        assert_equal 'application/citeproc+json', doi.response_content_type
      end

      def test_lookup_xml
        expected = '<?xml version="1.1><root><node>value</node>"'

        stub_request(:get, /https:\/\/dx.doi.org\/(.+)xml/)
          .to_return(body: 'This is redirecting', status: 301, headers: { 'location' => 'http://example.com/doixml' })
        stub_request(:get, 'http://example.com/doixml')
          .to_return(body: expected, status: 200, headers: { 'content-type' => 'application/unixref+xml' })

        # Lookup with curb
        doi = SimpleDOI::DOI.new '10.1234/efgh.xml'
        assert_nil doi.response_content_type, 'No content type should be set before lookup'
        assert_nil doi.body, 'No body should be set before lookup'

        doi.backend = 'curb'
        hash = doi.lookup_xml
        assert_equal expected, hash
        assert_equal 'application/unixref+xml', doi.response_content_type
        assert_equal expected, doi.body

        # Lookup with Net::HTTP
        doi = SimpleDOI::DOI.new '10.1234/efgh.xml'
        doi.backend = 'net/http'
        xml = doi.lookup_xml
        assert_equal expected, xml
        assert_equal 'application/unixref+xml', doi.response_content_type
        assert_equal expected, doi.body
      end

      def test_lookup_multi
        expected_xml = '<?xml version="1.1><root><node>value</node>"'

        stub_request(:get, /https:\/\/dx.doi.org\/(.+)xml/)
          .to_return(body: 'This is redirecting', status: 301, headers: { 'location' => 'https://example.com/doixml' })
        stub_request(:get, 'https://example.com/doixml')
          .to_return(body: expected_xml, status: 200, headers: { 'content-type' => 'application/unixref+xml' })

        expected_json = '{"prop1":"property", "prop2":"property"}'

        stub_request(:get, /https:\/\/dx.doi.org\/(.+)json/)
          .to_return(body: 'This is redirecting', status: 301, headers: { 'location' => 'http://example.com/doijson' })
        stub_request(:get, 'http://example.com/doijson')
          .to_return(body: expected_json, status: 200, headers: { 'content-type' => 'application/citeproc+json' })

        # Lookup with curb
        doi = SimpleDOI::DOI.new '10.1234/efgh.xml'
        assert_nil doi.response_content_type, 'No content type should be set before lookup'
        assert_nil doi.body, 'No body should be set before lookup'

        doi.backend = 'curb'
        # Lookup preferring xml
        body = doi.lookup([SimpleDOI::UNIXREF_XML, SimpleDOI::CITEPROC_JSON])
        assert_equal expected_xml, body
        assert_equal 'application/unixref+xml', doi.response_content_type
        assert_equal expected_xml, doi.body

        # No XML return, should get JSON
        doi = SimpleDOI::DOI.new '10.1234/efgh.json'
        assert_nil doi.response_content_type, 'No content type should be set before lookup'
        assert_nil doi.body, 'No body should be set before lookup'

        doi.backend = 'curb'
        # Lookup preferring xml, but actually get JSON
        body = doi.lookup([SimpleDOI::UNIXREF_XML, SimpleDOI::CITEPROC_JSON])
        assert_equal expected_json, body
        assert_equal 'application/citeproc+json', doi.response_content_type
        assert_equal expected_json, doi.body
      end

      def test_lookup_404
        stub_request(:get, 'https://dx.doi.org/10.1234%2Fnotexist')
          .to_return(body: 'DOI DOES NOT EXIST', status: 404, headers: { 'content-type' => 'text/html' })

        # Lookup with curb
        doi = SimpleDOI::DOI.new '10.1234/notexist'
        doi.backend = 'curb'
        assert_nil doi.lookup_json, 'Non-existing DOI should return nil'
        assert_nil doi.body
        assert_nil doi.response_content_type

        # Lookup with Net::HTTP
        doi = SimpleDOI::DOI.new '10.1234/notexist'
        doi.backend = 'net/http'
        assert_nil doi.lookup_json, 'Non-existing DOI should return nil'
        assert_nil doi.body
        assert_nil doi.response_content_type
      end

      def test_invalid_content_type_raises_exception
        stub_request(:get, /https:\/\/dx.doi.org\/(.+)/)
          .to_return(body: 'This is redirecting', status: 301, headers: { 'location' => 'http://example.com/doijson' })
        stub_request(:get, 'http://example.com/doijson')
          .to_return(body: 'This returns HTML', status: 200, headers: { 'content-type' => 'text/html' })

        doi = SimpleDOI::DOI.new '10.1234/abcd'
        doi.backend = 'curb'
        assert_raises SimpleDOI::InvalidResponseContentTypeError, 'An error should be raised when an unsupported content-type is returned' do
          doi.lookup
          puts doi.response_code
          puts doi.response_content_type

          puts doi.body
        end
      end

      def test_target_url
        stub_request(:get, 'https://dx.doi.org/10.1234%2Ftarget')
          .to_return(body: 'REDIRECT TO LOCATION', status: 301, headers: { 'location' => 'http://example.com/target' })
        stub_request(:get, 'https://dx.doi.org/10.1234%2Ftarget_notexist')
          .to_return(body: 'DOI DOES NOT EXIST', status: 404)

        doi = SimpleDOI::DOI.new '10.1234/target'
        doi.backend = 'curb'
        assert_equal 'http://example.com/target', doi.target_url, 'The target_url should match expected'

        doi = SimpleDOI::DOI.new '10.1234/target'
        doi.backend = 'net/http'
        assert_equal 'http://example.com/target', doi.target_url, 'The target_url should match expected'

        doi = SimpleDOI::DOI.new '10.1234/target_notexist'
        doi.backend = 'curb'
        assert_nil doi.target_url, 'Non-existent DOI should return nil'

        doi = SimpleDOI::DOI.new '10.1234/target_notexist'
        doi.backend = 'net/http'
        assert_nil doi.target_url, 'Non-existent DOI should return nil'
      end
    end
  end
end


lib/metadata_parser.rb
module SimpleDOI
  module MetadataParser
    # Metadata parser base class
    class Parser
      PROPERTIES = [
        :book_title,
        :book_series_title,
        :chapter_title,
        :isbn,
        :eisbn,
        :journal_title,
        :journal_isoabbrev_title,
        :issn,
        :eissn,
        :article_title,
        :conference_title,
        :contributors,
        :doi,
        :url,
        :publisher,
        :volume,
        :issue,
        :pagination,
        :publication_date
      ].freeze

      Contributor = Struct.new(:given_name, :surname, :contributor_role, :sequence)

      # Default readers for all PROPERTIES
      attr_reader :str, *PROPERTIES

      # Params:
      # +str+:: +String+ XML String
      def initialize(str)
        @str = str
      end

      def journal?
        raise NotImplementedError
      end

      def journal_article?
        raise NotImplementedError
      end

      def book?
        raise NotImplementedError
      end

      def book_series?
        raise NotImplementedError
      end

      def authors
        contributors.select {|c| c.contributor_role == 'author'}
      end

      def editors
        contributors.select {|c| c.contributor_role == 'editor'}
      end

      def contributors
        @contributors ||= []
      end

      def publisher
        # Return publisher name concatenated with place if available
        # If would throw a method error for :+ on publisher_name if nil, to dump nil for the whole thing
        publisher_name + (publisher_place ? "; #{publisher_place}" : "") rescue nil
      end

      def publisher_name
        raise NotImplementedError
      end

      def publisher_place
        raise NotImplementedError
      end

      # Return all properties as a Hash
      def to_h
        hash = {}
        PROPERTIES.each { |property| hash[property] = send property }
        # Array of Struct needs additional handling
        hash[:contributors] = @contributors.map(&:to_h)
        hash
      end
      alias :to_hash :to_h
    end
  end
end


lib/metadata_parser/unixref_xml_parser.rb
require 'nokogiri'
require_relative '../metadata_parser'

module SimpleDOI
  # Metadata parsers
  module MetadataParser
    # Basic metadata parser for Unixref+xml content types
    class UnixrefXMLParser < Parser
      XPATH_ROOT = '/doi_records/doi_record/crossref'.freeze

      # Params:
      # +str+:: +String+ XML String
      def initialize(str)
        super

        # Nokogiri will throw an exception on bad XML here
        @xml = Nokogiri::XML(@str) { |opts| opts.strict }
      end

      def book?
        !@xml.search("/#{XPATH_ROOT}/book/book_metadata").empty?
      end

      def book_series?
        !@xml.search("/#{XPATH_ROOT}/book/book_series_metadata").empty?
      end

      def conference_proceeding?
        !@xml.search("/#{XPATH_ROOT}/conference/proceedings_metadata").empty?
      end

      def journal_article?
        !@xml.search("/#{XPATH_ROOT}/journal/journal_article").empty?
      end

      def journal?
        !@xml.search("/#{XPATH_ROOT}/journal").empty? && !journal_article?
      end

      def book_title
        if book_series?
          @book_title ||= @xml.search("/#{XPATH_ROOT}/book/book_series_metadata/titles/title").first.text.strip rescue nil
        elsif conference_proceeding?
          @book_title ||= @xml.search("/#{XPATH_ROOT}/conference/proceedings_metadata/proceedings_title").first.text.strip rescue nil
        else
          @book_title ||= @xml.search("/#{XPATH_ROOT}/book/book_metadata/titles/title").first.text.strip rescue nil
        end
      end

      def book_series_title
        @book_series_title ||= @xml.search("/#{XPATH_ROOT}/book//series_metadata/titles/title").first.text.strip rescue nil
      end

      def conference_title
        @conference_title ||= @xml.search("/#{XPATH_ROOT}/conference//event_metadata/conference_name").first.text.strip rescue nil
      end

      def isbn
        # Usually includes media_type="print" but some vendors supply <isbn> with no attributes
        @isbn ||= @xml.search("/#{XPATH_ROOT}//isbn[not(@media_type) or @media_type=\"print\"]").first.text.strip rescue nil
      end

      def eisbn
        @eisbn ||= @xml.search("/#{XPATH_ROOT}//isbn[@media_type=\"electronic\"]").first.text.strip rescue nil
      end

      def journal_title
        @journal_title ||= @xml.search("/#{XPATH_ROOT}/journal/journal_metadata/full_title").first.text.strip rescue nil
      end

      def journal_isoabbrev_title
        @journal_isoabbrev_title ||= @xml.search("/#{XPATH_ROOT}/journal/journal_metadata/abbrev_title").first.text.strip rescue nil
      end

      def issn
        # Usually includes media_type="print" but some vendors supply <issn> with no attributes
        @issn ||= @xml.search("/#{XPATH_ROOT}//issn[not(@media_type) or @media_type=\"print\"]").first.text.strip rescue nil
      end

      def eissn
        @eissn ||= @xml.search("/#{XPATH_ROOT}//issn[@media_type=\"electronic\"]").first.text.strip rescue nil
      end

      def article_title
        if conference_proceeding?
          @article_title ||= @xml.search("/#{XPATH_ROOT}/conference/conference_paper/titles/title").first.text.strip rescue nil
        else
          @article_title ||= @xml.search("/#{XPATH_ROOT}/journal/journal_article/titles/title").first.text.strip rescue nil
        end
      end

      def contributors
        @contributors ||= (@xml.search(contributors_path).map.with_index(1) do |contributor, idx|
          Contributor.new(
            (contributor.search('./given_name').first.text.strip rescue nil),
            (contributor.search('./surname').first.text.strip rescue nil),
            (contributor.attr('contributor_role').strip rescue nil),
            idx
          )
        end)
      end

      def doi
        @doi ||= (@xml.search("#{doi_path}/doi").first.text.strip rescue nil) || (@xml.search("#{XPATH_ROOT}//doi_data/doi").first.text.strip rescue nil)
      end

      def url
        @url ||= (@xml.search("#{doi_path}/resource").first.text.strip rescue nil) || (@xml.search("#{XPATH_ROOT}//doi_data/resource").first.text.strip rescue nil)
      end

      def publisher_name
        @publisher_name ||= @xml.search("//publisher/publisher_name").first.text.strip rescue nil
      end

      def publisher_place
        @publisher_place ||= @xml.search("//publisher/publisher_place").first.text.strip rescue nil
      end

      def volume
        @volume ||= (@xml.search("#{volume_path}").first.text.strip rescue nil)
      end

      def issue
        @issue ||= (@xml.search("#{volume_issue_path}/issue").first.text.strip rescue nil)
      end

      def pagination
        @pagination ||= (@xml.search("//pages/first_page").first.text.strip + '-' + @xml.search('//pages/last_page').first.text.strip rescue nil)
      end

      def publication_date
        Date.new(publication_date_hash[:year], publication_date_hash[:month] || 1, publication_date_hash[:day] || 1) rescue nil
      end

      def publication_date_hash
        @publication_date_hash ||= {
          year: (@xml.search("#{publication_date_path}/year").first.text.strip.to_i rescue nil),
          month: (@xml.search("#{publication_date_path}/month").first.text.strip.to_i rescue nil),
          day: (@xml.search("#{publication_date_path}/day").first.text.strip.to_i rescue nil)
        }
      end

      protected

      def doi_path
        xpath = XPATH_ROOT
        if journal_article?
          xpath + '/journal/journal_article/doi_data'
        elsif journal?
          xpath + '/journal/journal_metadata/doi_data'
        elsif book?
          xpath + '/book/book_metadata/doi_data'
        elsif book_series?
          xpath + '/book/book_series_metadata/doi_data'
        else
          xpath + '//doi_data'
        end
      end

      def contributors_path
        xpath = XPATH_ROOT
        if journal_article?
          xpath + '/journal/journal_article/contributors/person_name'
        elsif book?
          xpath + '/book/book_metadata/contributors/person_name'
        elsif book_series?
          xpath + '/book/book_series_metadata/contributors/person_name'
        elsif conference_proceeding?
          xpath + '/conference/conference_paper/contributors/person_name'
        else
          xpath + '//contributors'
        end
      end

      def volume_issue_path
        xpath = XPATH_ROOT
        if journal_article?
          xpath + '/journal/journal_issue'
        elsif book?
          xpath + '/book/book_metadata'
        elsif book_series?
          xpath + '/book/book_series_metadata'
        elsif conference_proceeding?
          xpath + '/conference/proceedings_metadata'
        else
          '//'
        end
      end

      def volume_path
        xpath = volume_issue_path
        if journal_article?
          xpath + '/journal_volume/volume'
        elsif book? || book_series?
          xpath + '/volume'
        else
          '//volume'
        end
      end

      def publication_date_path
        '//publication_date'
      end
    end
  end
end


lib/simple_doi/simple_doi.rb
require 'json'
require 'cgi'
require 'uri'

# A simple DOI handling library
# Capable of validating DOIs and retrieving metadata
module SimpleDOI
  CITEPROC_JSON = 'application/vnd.citationstyles.csl+json'
  UNIXREF_XML = 'application/vnd.crossref.unixref+xml'

  # Regex pattern NOT anchored with ^$
  # Validation will add anchors
  # DOI_PATTERN = '\b(10\.[\d]{4,}(?:\.[\d]+)*\/(?:(?!["&\'<>])[[:graph:]])+)\b'
  DOI_PATTERN = '\b(10\.[\d]+(?:\.[\d]+)*\/(?:(?!["&\'])[[:graph:]])+)\b?'

  STRIP_PATTERNS = [
    /\/(abstract|asset|issuetoc).*$/,
    /\.pdf$/,
    /\/(standard|pdf\/standard|fulltext\.html|pdf|full|dynaTraceMonitor|references|issues|full)$/,
    /\/cite\/[a-z]+$/,
    /;jsessionid.+$/
  ].freeze

  # Convenience methods to call directly on module
  def valid?(doi)
    !!(strip(doi) =~ Regexp.new('^' + DOI_PATTERN + '$'))
  end

  # Attempt to extract all DOIs from string
  # Returns an array of SimpleDOI::DOI objects
  def extract(string)
    string = CGI.unescape string
    list = (string.scan(Regexp.new(DOI_PATTERN))).flatten
    # Calls strip() to make sure nothing like ?nosfx=y remains
    list.map! { |doi| strip doi }

    dois = []
    list.each { |doi| dois.push(SimpleDOI::DOI.new(doi)) if valid?(doi) }
    dois
  end

  # Strip off dx.doi.org extra stuff
  def strip(string)
    # Strip off leading DOI: if present
    string.sub!(/^doi:/i, '')

    # Strip off doi.org URL prefix and ?nosfx
    string.sub!(Regexp.new('^https?://' + DOI::LOOKUP_URL_DOMAIN + '/'), '')

    # Strip query string
    # Have to make sure it actually looks like a query string with key=value
    # because ? are not expressly prohibited in a DOI.
    string.sub!(/\?[^=]+=.*/, '')

    # Strip trailing /
    string.sub!(/\/$/, '')

    # Aggressively strip known URL patterns which tend not to be part
    STRIP_PATTERNS.each { |pattern| string.sub!(pattern, '') }

    string
  end
  module_function :valid?, :extract, :strip

  # Represents a DOI object capable of being resolved to a target URL or
  # full metadata retrieved.
  class DOI
    include SimpleDOI

    LOOKUP_URL_DOMAIN = 'dx.doi.org'

    BACKENDS = %w(curb net/http)

    REDIR_CODES = [301, 302, 303]

    attr_reader :body, :response_content_type, :response_code, :backend, :doi

    def initialize(doi)
      # Prefer curb if available, otherise Net::HTTP
      @backend = 'curb' if defined?(Curl)
      @backend ||= 'net/http' if defined?(Net::HTTP)

      raise ArgumentError, "Supplied string does not appear to be a valid DOI: #{doi}" unless valid?(doi)
      @doi = doi
      @target_url = nil
    end

    def to_s
      @doi
    end

    def backend=(backend)
      if BACKENDS.include? backend
        @backend = backend
        require backend
      else
        raise ArgumentError, "Only '#{BACKENDS.join("', '")}' are supported"
      end
    end

    def prefix
      to_s.split('/').first
    end

    def suffix
      to_s.split('/', 2).last
    end

    # Return an ISSN string if the DOI suffix matches:
    # j.1234-567X
    # issn.1234-567X
    # 1234-567X.otherstuff
    # ScienceDirect prefixes with S
    # S1234-567X(otherstuff)
    # (ISSN)1234-567X.otherstuff
    def issn
      suffix.scan(/^(?:(?:j\.|issn\.)|S|\(ISSN\))?(\d{4}-\d{3}[\dx])\b/i).flatten.first
    end

    # Perform a lookup for DOI metadata, requesting content types
    # specified in accept
    # Returns: +String+ HTTP response body
    # Params:
    # * +accept+:: +String|Array+ Requested Accept: content types may be single string or array
    def lookup(accept = CITEPROC_JSON)
      raise NoBackendError unless @backend

      # Coerce accept to a 1D array
      accept = [accept].flatten
      @body = nil
      @response_content_type = nil
      @response_code = nil

      case @backend
      when 'curb'
        client = Curl::Easy.new(url) do |request|
          request.headers['Accept'] = accept.join(', ')
          request.follow_location = true
        end
        client.perform
        @response_code = client.response_code.to_i
        if @response_code == 200
          @body = client.body_str
          # Curl returns all headers a string block
          # We only need Content-Type
          # The *last* Content-Type from redirection is the relevant one
          @response_content_type = client.header_str.scan(/content-type:\s+(\S+)/i).flatten.pop rescue nil
        end
      when 'net/http'
        uri = URI.parse url

        response = net_http_response_redirect(uri, accept)
        @response_code = response.code.to_i
        if @response_code == 200
          @body = response.body
          @response_content_type = response['content-type']
        end
      end

      # Raise an error if the document ultimately retrieved doesn't have a JSON or XML
      # content type.
      # We've seen some JSTOR documents redirect 3+ times and return a 200 with HTML
      # instead of the requested metadata - basically ignored content negotiation.
      if @response_code == 200 && @response_content_type !~ /json|xml/
        raise InvalidResponseContentTypeError
      end
      @body
    end

    def lookup_xml
      lookup UNIXREF_XML
    end

    def lookup_json
      resp = lookup CITEPROC_JSON
      JSON.parse(resp) unless resp.nil?
    end

    def self.lookup_url_root
      "https://#{LOOKUP_URL_DOMAIN}/"
    end

    # Get the redirection target URL or return it if already retrieved
    def target_url
      return @target_url unless @target_url.nil?

      # Perform an HTTP request but don't follow the first redirect
      case @backend
      when 'curb'
        client = Curl::Easy.new(url) do |request|
          request.max_redirects = 0
          request.follow_location = false
        end
        client.perform
        location = client.header_str.scan(/location:\s+(\S+)/i).flatten.first if REDIR_CODES.include? client.response_code.to_i
      when 'net/http'
        uri = URI.parse url
        request_use_ssl = uri.scheme == 'https'
        client = Net::HTTP.new uri.host, (request_use_ssl ? Net::HTTP.https_default_port : Net::HTTP.http_default_port)
        client.use_ssl = request_use_ssl
        request = Net::HTTP::Get.new uri.request_uri
        response = client.request request
        location = response['location'] if REDIR_CODES.map(&:to_s).include? response.code
      end
      @target_url = location
    end

    def url
      "#{self.class.lookup_url_root}#{CGI.escape(to_s)}"
    end

    protected

    def net_http_response_redirect(uri, type)
      # Net::HTTP require different initial setup for https vs http
      request_use_ssl = uri.scheme == 'https'
      client = Net::HTTP.new uri.host, (request_use_ssl ? Net::HTTP.https_default_port : Net::HTTP.http_default_port)
      client.use_ssl = request_use_ssl
      request = Net::HTTP::Get.new uri.request_uri
      request['Accept'] = type
      response = client.request request

      # Recurse until done redirecting
      if REDIR_CODES.map(&:to_s).include? response.code
        net_http_response_redirect(URI.parse(response['location']), type)
      else
        response
      end
    end
  end

  class DOIError < StandardError; end
  class InvalidResponseContentTypeError < DOIError; end

  # Raised when no supported HTTP client backend was found in the constant space
  class NoBackendError < DOIError
    def initialize(msg = "No supported backend was found. You must require either 'curb' or 'net/http'")
      super msg
    end
  end
end
